qiskit_experiments/framework/composite/batch_experiment.py

```diff
@@ -127,3 +127,69 @@ def _remap_qubits(self, circuit, qubit_mapping):
         new_circuit.metadata = circuit.metadata
         new_circuit.append(circuit, qubit_mapping, list(range(num_clbits)))
         return new_circuit
+
+    def _run_jobs(self, circuits: List[QuantumCircuit], **run_options) -> List[Job]:
+        if len(circuits) == 0:
+            return []
+
+        if self.experiment_options.separate_jobs:
+            try:
+                is_first_time = False
+                # The local metadata is a truncation of the original composite metadata,
+                # such that the current experiment (self) is at the head of the metadata.
+                # In the root experiment (at the beginning of the reucrsion) none of the circuits
+                # contains the local_metadata attribute; after the first iteration they all do.
+                if not hasattr(circuits[0], "local_metadata"):
+                    is_first_time = True
+                    for circ in circuits:
+                        circ.local_metadata = circ.metadata
```

Comment:
I don't think dynamically adding attribute like this is right approach. For example, if we decide to set `__slots__` to `QuantumCircuit`, this will no longer work. How local_metadata attribute is consumed by the job? Seems like this is only used to sort flatten circuits by sub experiment index. I think you can separately store list of metadata, and do something like
```python
circs_by_subexps = {}
for circ, metadata in zip(circuits, metadata_list):
    ...
    # sort by component index, add truncated metadata
```